### ISSUE TITLE
Memory leak when training models

### ISSUE THREAD

**Issue by brendanhasz** [brendanhasz/probflow#12]
Especially noticable when training on large datasets.  Not sure if it's on ProbFlow's end or TensorFlow's (but almost surely ProbFlow's haha).

Training on 1M rows from the [taxi trip dataset](https://console.cloud.google.com/marketplace/details/city-of-new-york/nyc-tlc-trips) w/ TF GPU caused a leak of around 50MB/epoch :grimacing:  Was even worse w/ TF CPU

Likely is a problem with [DataGenerator](https://probflow.readthedocs.io/en/latest/api_data.html).  50MB with that test was about exactly the size of the x+y dataset, so DataGenerator is probably making copies when shuffling (instead of creating views) and somehow holding on to them ad infinitum.

Also was using MonitorMetric callback, that might be the problem instead (or the corresponding Model.metric / metric function).

Also some preliminary poking around w/ pympler found that a bunch of large pandas Series were being created, but wasn't training using pandas arrays, so don't know where those were coming from.

A few things to test:

* Does training w/o shuffling solve the problem?
* Does training w/o MonitorMetric callback solve the problem?

Simplified code causing the problem:
```python
import numpy as np
import probflow as pf

# Data
x_train = np.random.randn(1000000, 8).astype('float32')
y_train = np.random.randn(1000000, 1).astype('float32')
x_val =   np.random.randn( 100000, 8).astype('float32')
y_val =   np.random.randn( 100000, 1).astype('float32')

# Callbacks
monitor_elbo = pf.MonitorELBO()
monitor_mae = pf.MonitorMetric('mae', x_val, y_val)
lr_scheduler = pf.LearningRateScheduler(lambda e: 2e-4-2e-6*e)
callbacks = [monitor_elbo, monitor_mae, lr_scheduler]

# Model
model = pf.DenseRegression([7, 256, 128, 64, 32, 1])

# Train
model.fit(x_train, y_train,
          epochs=100,
          batch_size=1024,
          callbacks=callbacks)
```
